### PR TITLE
Support Traefik ForwardAuth without a 401 handler

### LINKED ISSUE #1015: [Feature Request] Allow auth endpoint to serve sign_in pages as well
Mentioned at https://github.com/oauth2-proxy/oauth2-proxy/pull/762#issuecomment-768360213 by @pcneo83

> Is there any chance on reconsidering the missing functionality that hasn't been captured in #957 please? While `X-Forwarded-*` headers are now supported to make the redirects capture the path correctly, this does not work on the initial unauthenticated requests with Traefik. The dependency on `errors` middleware to capture 401 status and then initiate `/oauth2/sign_in` breaks the signin redirect in the browser (when --skip-provider-button is true).
> 
> The `--auth-endpoint-sign-in` option suggested in [#762 (comment)](https://github.com/oauth2-proxy/oauth2-proxy/pull/762#issuecomment-688440183) and implemented in this PR is the only missing functionality and would greatly help in using Traefik FowardAuth using a single endpoint `oauth2/auth` and will also fix
> 
> * --skip-provider-button redirect failure in the browser
> * missing unauthenticated request URI path in the redirect

This should be implemented, as it is noted that when `--skip-provider-button` is set, the redirect doesn't work.

### PR DESCRIPTION
**Pull request by pcneo83**
Adds support for Traefik ForwardAuth middleware without the need to handle 401 unauthorized response using an `errors` middleware

- overload `/oauth2/auth` endpoint with a QS boolean parameter to start Oauth2 process when required
- fixes browser redirect issues with 401 response when `--skip-provider-button` is enabled

Fixes #1015 

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: oauthproxy.go
Comment (JoelSpeed) [inline oauthproxy.go:854]: What would happen if Traefik received a 302 code in the response here? Could it follow the redirect?

File: oauthproxy_test.go
Comment (JoelSpeed) [inline oauthproxy_test.go:2952]: Why does this end up as a forbidden response? Do we have a test case for a successfuly redirect_signin, ie it would render a page and return and ok status?

**Review by JoelSpeed — COMMENTED** [review]
This is an interesting idea, left a couple of questions. I'm not hugely keen on this overloading of the authonly endpoint, it would be better if we could change the behaviour to actually redirect rather than rendering the page, is that possible?

File: oauthproxy_test.go
Comment (pcneo83) [reply oauthproxy_test.go:2952]: I did not know how to pass a valid token for the `SignInPage` call to succeed, so made the test name indicate an invalid token is being passed. I can have a look to add a successful render test case but would appreciate if you could point to an existing test that passes a valid token, will certainly make it easier for me

File: oauthproxy.go
Comment (pcneo83) [reply oauthproxy.go:854]: I have looked again at Traefik ForwardAuth middleware docs, and seems to indicate that it will `return the response from the authentication server to the client`, which perhaps indicates that a 302 redirect might work. I'll give it a try and get back to you on that one

Comment (pcneo83) [reply oauthproxy.go]: I have added a redirect option now and left the commented out code for rendering options as well, interested to hear your thoughts

File: oauthproxy_test.go
Comment (pcneo83) [reply oauthproxy_test.go]: With the redirect option, the test checks for `302` but will add more tests if there is an agreement on the approach please

File: oauthproxy.go
Comment (JoelSpeed) [inline oauthproxy.go:870]: We don't need this to be logged so please remove before we merge

File: oauthproxy.go
Comment (JoelSpeed) [inline oauthproxy.go:872]: This should be the sign in path not sign out path

File: oauthproxy.go
Comment (JoelSpeed) [inline oauthproxy.go:875]: This will be handled by the sign in page, so I don't think we need to duplicate the logic here

File: oauthproxy.go
Comment (JoelSpeed) [inline oauthproxy.go:843]: I don't think we would want this logged in production, please remove before we merge

**Review by JoelSpeed — COMMENTED** [review]
I definitely prefer the redirect option as it's implemented now. Did you test this and it worked as expected? 

@NickMeves what do you think to this approach to allow the flexibility for traefik?

File: oauthproxy.go
Comment (NickMeves) [inline oauthproxy.go:1248]: Does this potentially make sense as an explicit URL to redirect to in unauthenticated `/oauth2/auth` cases? Gives more flexibility beyond Traefik and eliminates all the signin URL determination logic above.

Mirrors the setup nginx & kubernetes ingress does where they set the explicit signin-url as an annotation.

**Comment by JoelSpeed** [comment]
Just thinking about this, have you tried to use a static upstream as was introduced in #265?

The default behaviour for accessing any endpoint that is not authenticated is to return the sign in page, so I think you may be able to configure this for your needs

**Comment by pcneo83** [comment]
> I definitely prefer the redirect option as it's implemented now. Did you test this and it worked as expected?
Sounds good, this has been tested to work in our pre-prod environments but hasn't been pushed to prod yet. If the approach is ok, I'll look into adding more test cases if we are missing any. 

redirect option makes sense as its handling the case at a HTTP level, however, it would also mean duplicating the code to keep track of the original request, before its redirected to https://oauth2-proxy-service.example.com/oauth2/(sign_in|start). If not, the original request is lost and the sign_in|start logic will only redirect back to https://oauth2-proxy-service.example.com once OAuth2 completes

**Comment by pcneo83** [comment]
> Just thinking about this, have you tried to use a static upstream as was introduced in #265?
> 
> The default behaviour for accessing any endpoint that is not authenticated is to return the sign in page, so I think you may be able to configure this for your needs

I might have misunderstood how to use the _upstream_ functionality but I did try with `--upstream="static://202"` with the following set to Traefik ForwardAuth address:
- https://oauth2-proxy-service.example.com/oauth2/auth - gives 401 unauthorized for unauthenticated sessions
- https://oauth2-proxy-service.example.com -

File: oauthproxy.go
Comment (pcneo83) [reply oauthproxy.go:861]: Perhaps I'm misunderstanding, but the request flow would be
- https://a-service.example/com/testing
- Traefik ForwardAuth sends to https://oauth2-proxy.example.com/oauth2/auth?redirect_signin=true
- oauth2-proxy checks the session is unauthenticated and redirects back to request URL which will be https://oauth2-proxy.example.com/oauth2/(sign_in|start).

So without the duplication logic of retrieving the original request (https://service-a.example.com) and placing it in an `rd` QS param, the original request is lost

File: oauthproxy.go
Comment (pcneo83) [reply oauthproxy.go]: removed

File: oauthproxy.go
Comment (pcneo83) [reply oauthproxy.go]: fixed

File: oauthproxy.go
Comment (pcneo83) [reply oauthproxy.go]: Have removed the `SkipProviderButton` duplicated check now, as you pointed out the `SignIn` method is already doing this

File: oauthproxy.go
Comment (pcneo83) [reply oauthproxy.go]: removed

File: oauthproxy.go
Comment (pcneo83) [reply oauthproxy.go:1231]: I have switched the redirect to `/oauth2/sign_in` endpoint as the `SignIn` method seems to handle the both SignInPage or OauthStart options depending upon the `SkipProviderButton` value.

File: docs/docs/configuration/overview.md
Comment (linuxgemini) [inline docs/docs/configuration/overview.md:506]: This example does not utilize the File Dynamic Configuration Reference which is shown on Traefik Docs: https://doc.traefik.io/traefik/reference/dynamic-configuration/file/

Different config styles documented for the same software target can create confusion.

**Review by linuxgemini — COMMENTED** [review]
This example should be re-written with the File Dynamic Configuration Reference in mind.

The rules can be handled better as well.

**Comment by NickMeves** [comment]
> > I definitely prefer the redirect option as it's implemented now. Did you test this and it worked as expected?
> > Sounds good, this has been tested to work in our pre-prod environments but hasn't been pushed to prod yet. If the approach is ok, I'll look into adding more test cases if we are missing any.
> 
> redirect option makes sense as its handling the case at a HTTP level, however, it would also mean duplicating the code to keep track of the original request, before its redirected to [https://oauth2-proxy-service.example.com/oauth2/(sign_in|start)](https://oauth2-proxy-service.example.com/oauth2/(sign_in%7Cstart)). If not, the original request is lost and the sign_in|start logic will only redirect back to https://oauth2-proxy-service.example.com once OAuth2 completes

Take a look at this writeup on the nginx ingress configuration of this: https://kubernetes.github.io/ingress-nginx/examples/auth/oauth-external-auth/

Particularly here:

```
metadata:
  name: application
  annotations:
    nginx.ingress.kubernetes.io/auth-url: "https://$host/oauth2/auth"
    nginx.ingress.kubernetes.io/auth-signin: "https://$host/oauth2/start?rd=$escaped_request_uri"
```

Based on a failure case, could you set a parameter on the `/oauth2/auth` endpoint and you construct your `https://$host/oauth2/start?rd=$escaped_request_uri` parameter from your Traefik config?

I lean toward making this is broad as possible so other systems or use cases could use it and not tailoring the solution purely to traefik ease of use since it likely only a small minority of the reverse proxies that sit in front of OAuth2 Proxy.

**Comment by pcneo83** [comment]
> Based on a failure case, could you set a parameter on the `/oauth2/auth` endpoint and you construct your `https://$host/oauth2/start?rd=$escaped_request_uri` parameter from your Traefik config?

If you mean to update Traefik ForwardAuth middleware address endpoint to automatically set the `rd` parameter, its unfortunately not possible due to Traefik configuration being very static and its lack of support for dynamic variables.

We are currently setting the middleware to below and configuring any service route path that needs to be protected by authentication with the below middleware configuration. Even though each service and its path can be configured with the middleware, due to the lack of dynamic variable support, its not possible to use and replace variables

```
forwardAuth:
    address: https://oauth2-proxy-service.example.com/oauth2/auth?redirect_sign=true
```

**Comment by linuxgemini** [comment]
I have digged around and have seen that Traefik doesn't pass through redirect responses of error middlewares correctly and mashes up the 302 response with a 401 with the "Found" page and with a Location header.



This is likely a Traefik bug or an implementation choice.

I have opened https://github.com/traefik/traefik/issues/7872 to investigate.

File: docs/docs/configuration/overview.md
Comment (pcneo83) [reply docs/docs/configuration/overview.md]: It was easier to use my existing Kubernetes provider configuration as the examples at the time, but have now updated to use Dynamic File Provider. Is it worth also keeping the Kubernetes Provider CRD example here or somewhere else as its not always straightforward to switch between different configuration providers easily?

**Comment by pcneo83** [comment]
@NickMeves, @JoelSpeed appreciate any further thoughts/concerns on this please? Have resolved the points raised code wise and have also updated the tests.

File: oauthproxy.go
Comment (NickMeves) [inline oauthproxy.go:842]: Maybe flip this to:

```suggestion
		if !authOnlyRedirectSignIn(req) {
				http.Error(rw, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
				return
		}
```

So we can eliminate the deep nesting for all that custom logic below

File: oauthproxy.go
Comment (NickMeves) [inline oauthproxy.go:849]:
```suggestion
			// Original request URI is captured and added using `rd` parameter in the redirect
```

File: oauthproxy.go
Comment (NickMeves) [inline oauthproxy.go:854]: Make a reminder note:

https://github.com/oauth2-proxy/oauth2-proxy/pull/1029 will likely merge before this. When you rebase, you'll need to update to use the new style of error pages.

File: oauthproxy_test.go
Comment (NickMeves) [inline oauthproxy_test.go:2942]:
```suggestion
func Test_authOnlyRedirectSignIn(t *testing.T) {
```

(to align to the private naming in the package)

File: oauthproxy.go
Comment (NickMeves) [inline oauthproxy.go:851]: `getAppRedirect` has a lot of logic paths to figure out where the app redirect goes to -- In traefik will this come from the `X-Forwarded-(Proto|Host|Uri)` headers?

If so and we then map them to `rd` on the SignIn handler -- that is acceptable (or is traffic still sending the same X headers to the sign in..?)

FYI - `rd` takes priority over everything else -- so this logic seems good to me. I just wanted to talk it out in case there's unintentional mixing of logic.

File: oauthproxy.go
Comment (NickMeves) [inline oauthproxy.go:860]: Do you need the full URL here?

E.g. -- what happens if this is used in a system hitting this from a subrequest auth check where the OAuth2-Proxy server is hosted on a different subdomain then the requested resource.

In that case, the redirect to the relative path will be at the application domain - not the separate oauth2-proxy server. (this separate central oauth2-proxy auth server hit via subrequests on a different subdomain is a common architecture deployment).

Comment (NickMeves) [reply oauthproxy.go:860]: And then how do we accurately know the appropriate `Proto/Host` details to set? If this endpoint is hit via subrequests, it might not have `X-Forwarded` details or be on different internal domain (and therefore different HTTP Host header).

For instance, when I use this architecture, I do subrequests to this endpoint over the Kubernetes service URL directly (e.g. `deployment.namespace.svc.cluster.local`) to avoid the extra hop of going though the Ingress layer internally.

The views the `/oauth2/auth` endpoint has of its `requestutil.GetRequestHost(req)` for those requests are not what is accessible externally.

Comment (NickMeves) [reply oauthproxy.go:860]: Upon further reflection -- this problem will cause you to get misleading results for the `p.getAppRedirect(req)` call above to try to set an `rd` querystring as well 🤔

File: docs/docs/features/endpoints.md
Comment (NickMeves) [inline docs/docs/features/endpoints.md:16]: I'm not sure if we want to document all the querystring available on each endpoint in this page (in order to avoid bloat).

This wouldn't even be the first supported querystring param on this endpoint (it has `allowed_groups` as well). Other endpoints don't document their various querystrings or headers on this page (but I do recall I made an issue to figure out a spot for us to document all supported querystrings and headers for all endpoints in a good place).

File: docs/docs/features/endpoints.md
Comment (NickMeves) [inline docs/docs/features/endpoints.md:15]: So the use cases for `/oauth2/auth` 202/401 (and 403 too now I think?) have grown well beyond these two use cases. (Kubenetes ingress annotations, nginx, traefik, direct auth checks from applications, etc).

I think we should simplify this endpoint documentation and not call out specific use cases anymore (just mention the generic utility of subrequest auth checks).

File: oauthproxy.go
Comment (NickMeves) [reply oauthproxy.go:860]: I think we might need to revisit `redirect_signin` not being a bool, but rather being an explicit URL to the sign in endpoint with the proto/host/path all set in your configuration.

File: oauthproxy.go
Comment (pcneo83) [reply oauthproxy.go]: updated

File: oauthproxy.go
Comment (pcneo83) [reply oauthproxy.go]: updated

File: oauthproxy.go
Comment (pcneo83) [reply oauthproxy.go]: updated

File: docs/docs/features/endpoints.md
Comment (pcneo83) [reply docs/docs/features/endpoints.md]: removed the additional query string parameter doc

File: docs/docs/features/endpoints.md
Comment (pcneo83) [reply docs/docs/features/endpoints.md]: Have removed the Nginx and Traefik specific statements and reworded. Please do suggest if you would like me to rephrase

File: oauthproxy.go
Comment (pcneo83) [reply oauthproxy.go]: Yes, Traefik uses the `X-Forwarded-(Proto|Host|Uri)` headers to provide the request URL information. The `getAppRedirect` seems to handle all the various cases nicely and the usage of `rd` parameter seemed like the best case here to preserve the original request URL

The flow is:
- original unauthenticated request comes in to `https://a-service.example.com/`
- Traefik ForwardAuth middleware forwards this request to `https://oauth2-proxy.example.com/oauth2/auth?redirect_signin=true`
- As per new logic, the original request URL is captured in `rd=https://a-service.example.com/` and added as a QS to `/oauth2/sign_in` redirect path with a 302 redirect returned to the client

File: oauthproxy.go
Comment (pcneo83) [reply oauthproxy.go]: Apologies but I quite don't understand by what you mean by `subrequests` but perhaps an example would help me. However, interesting that you mention bypassing ingress and the redirect logic depends upon ingress configuration.

> For instance, when I use this architecture, I do subrequests to this endpoint over the Kubernetes service URL directly (e.g. deployment.namespace.svc.cluster.local) to avoid the extra hop of going though the Ingress layer internally

Trying to understand how this looks for real requests, an end to end example of what the initial request is and how you bypass the ingress and make the `/oauth2/auth` request to the kubernetes internal endpoints would be useful. We do have to note that since `redirect_signin` is a boolean flag (for now), you could still switch off this behaviour by setting it to false when using internal endpoints?

> I think we might need to revisit `redirect_signin` not being a bool, but rather being an explicit URL to the sign in endpoint with the proto/host/path all set in your configuration.
I possibly didn't get your point correctly but any explicit configuration will not be possible to set when using ingress like Traefik

**Comment by pcneo83** [comment]
> > https://oauth2-proxy-service.example.com - redirects to /oauth2/sign_in or /oauth2/start depending on skip option set, but the original request gets a 404. Seem like oauth2-proxy is serving the request rather than the service's backend
> 
> Could you elaborate on this? I just tried out the static endpoint myself, if not logged in, accessing the root path of the proxy, I get the sign_in page rendered and a 403 response returned. If I am logged in, it returns a 202 exactly the same as the `/oauth2/auth` endpoint normally does.
> 
> As I understand it, that should satisfy Traefiks requirements
> 
> When you say OAuth2 Proxy is serving the request? Is that post login? That seems to suggest to me a misconfiguration on Traefik's side rather than an issue on OAuth2 Proxy

@JoelSpeed I can confirm `upstream=static://202` configuration works for the Traefik use case.

 I have tested this multiple times prior to v7 release and from my last message after the v7 release and yes it failed to redirect to the origin request URL post authentication. I believe the mis-configuration in my last testing was in oauth2-proxy where I might have not set `--reverse-proxy` to `true`. This meant that the `X-Forwarded-*` headers were not consumed and that resulted in an incorrect redirect to oauth2-proxy rather than the original URL post login

**Comment by JoelSpeed** [comment]
Ok cool, glad that it is working now! Can we get this PR updated to update the docs and drop the code changes as they are unnecessary right?

**Comment by JoelSpeed** [comment]
Docs changes look good as I understand them, are these also relevant for the 7.0.x and 6.1.x docs branches? Do we need to backport them?

You may want to add a changelog entry as well but I'll leave that up to you,

**Comment by pcneo83** [comment]
> Docs changes look good as I understand them, are these also relevant for the 7.0.x and 6.1.x docs branches? Do we need to backport them?

I have just tested with v6.1.1 version and this solution does not seem to work. It works with v7.0.0 however.

I'll update the changelog and push, thanks

**Comment by JoelSpeed** [comment]
Given that this works with v7, would you mind copying the changes into https://github.com/oauth2-proxy/oauth2-proxy/blob/master/docs/versioned_docs/version-7.0.x/configuration/overview.md which is for the v7.0.x release?

Also, there seems to be a changelog conflict, might need a rebase?

**Comment by pcneo83** [comment]
Missed the versioned docs, all updated now

**Comment by JoelSpeed** [comment]
Sorry, to keep the changes going forward, they need to be in the main docs folder (where you had it originally) and the versioned ones, could you re-add it to the other file please?

**Comment by pcneo83** [comment]
ah, my mistake, didn't realise both places need to be updated. I have updated both the docs now

File: docs/versioned_docs/version-7.0.x/configuration/overview.md
Comment (linuxgemini) [inline docs/versioned_docs/version-7.0.x/configuration/overview.md:462]: Where's `oauth.example.com/` handled?

Comment (pcneo83) [reply docs/versioned_docs/version-7.0.x/configuration/overview.md]: 🤦 , looks like I missed it during the refactor, added it now, thanks

**Review by JoelSpeed — APPROVED** [review]
I think this is good, @linuxgemini were you happy with the changes?

**Comment by linuxgemini** [comment]
> I think this is good, @linuxgemini were you happy with the changes?

@JoelSpeed LGTM as well

**Pull request merged**
